### PR TITLE
Maps few of default android configurations to environments

### PR DESCRIPTION
**Pull request by meghfossa**
**Opening this for early feedback.** 

# Overview

Maps few of the default android configurations to environment.

## Acceptance criteria

- Excludes development, and testing configurations associated with the android gradle build. 

## Testing plan

- I tested against building android project (Android Lollipop). My configurations were, 

**build.gradle**
```
buildscript {
    ext.kotlin_version = '1.5.21'
    repositories {
        google()
        mavenCentral()
    }
    dependencies {
        classpath 'com.android.tools.build:gradle:7.0.0'
        classpath "org.jetbrains.kotlin:kotlin-gradle-plugin:$kotlin_version"
    }
}
task clean(type: Delete) {
    delete rootProject.buildDir
}
```

**build.gradle**
```
plugins {
    id 'com.android.application'
    id 'kotlin-android'
}

android {
    compileSdk 31

    defaultConfig {
        applicationId "com.example.myapplication"
        minSdk 21
        targetSdk 21
        versionCode 1
        versionName "1.0"

        testInstrumentationRunner "androidx.test.runner.AndroidJUnitRunner"
    }

    buildTypes {
        release {
            minifyEnabled false
            proguardFiles getDefaultProguardFile('proguard-android-optimize.txt'), 'proguard-rules.pro'
        }
    }
    compileOptions {
        sourceCompatibility JavaVersion.VERSION_1_8
        targetCompatibility JavaVersion.VERSION_1_8
    }
    kotlinOptions {
        jvmTarget = '1.8'
    }
}

dependencies {
    implementation 'androidx.core:core-ktx:1.3.2'
    implementation 'androidx.appcompat:appcompat:1.2.0'
    implementation 'com.google.android.material:material:1.3.0'
    implementation 'androidx.preference:preference:1.1.1'
    testImplementation 'junit:junit:4.+'
    androidTestImplementation 'androidx.test.ext:junit:1.1.2'
    androidTestImplementation 'androidx.test.espresso:espresso-core:3.3.0'
}
```

## Risks

This is high risk PR, and can have ramification such as, failure to detect runtime dependency, if the project uses configuration name same, as default configuration generated by android build.

I intent to keep this PR open, and performing following analysis prior to closing
- [x] Documentation on how default configurations are generated
- [ ] Gradle script to identify the which plugins are specified in top level build.gradle, and making classification of android configuration contingent, if and only if, android build, and kotlin plugins exist
- [x] Potential impact on existing user's dependency count

## References

Closes https://github.com/fossas/team-analysis/issues/696

## Checklist

- [x] I added tests for this PR's change (or confirmed tests are not viable).
- [x] If this PR introduced a user-visible change, I added documentation into `docs/`.
- [x] I updated `Changelog.md` if this change is externally facing. If this PR did not mark a release, I added my changes into an `# Unreleased` section at the top.
- [x] I linked this PR to any referenced GitHub issues, if they exist.

### CODE REVIEW

File: test/Android/UtilSpec.hs [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+  , "implementation"
+  , "coreLibraryDesugaring"
+  , "releaseRuntimeOnlyDependenciesMetadata"
+  , "archives"
+  ]
+
+spec :: Spec
+spec = do
+  describe "isDefaultAndroidDevConfig" $ do
+    forM_ defaultAndroidDevConfigs $ \candidate -> do
```
Comment (scruffystuffs) [inline test/Android/UtilSpec.hs:40]: [nit]: I'm pretty sure you can use `for_` here instead of `forM_`.  I think the "M" in `forM_` means "monad".

**Review by cnr — APPROVED** [review]
looks structurally sound, risk mitigation plan looks good, and the dev/test classifications look reasonable to me at first glance

something from the risk mitigation plan I think might not be necessary, and might cause other edge cases:

> Gradle script to identify the which plugins are specified in top level build.gradle, and making classification of android configuration contingent, if and only if, android build, and kotlin plugins exist

In particular, the android gradle plugin doesn't need to be declared in the top-level build.gradle, and can for example be brought in through other first-party gradle plugins in subprojects (which I've seen in gradle configs of $largecustomer)

I'd be shocked to learn that the listed dev/test configuration names are being used in non-android projects for production dependencies, given that most of them are prefixed by `test` or `debug` or `lint`. If you do some digging and find out otherwise, we can revisit this bullet point

**Comment by meghfossa** [comment]
Leaving some of the analysis here, for posterity. 

After discussing with the team, we've decided to move forward with the approach. From risk mitigation analyses, conclusions were following:

(1) We cannot distinguish from analyses submitted between CLIv1, and CLIv2, and such we cannot infer the consequential impact. 

(2) Regardless of (1), performing few of snapshot queries (not exhaustive), common dependency originating development and test configurations were in magnitudes of 100s within last few months. 

Lastly, this will be documented in docs and change-log, and intentionally communicated to SEs (for potential drop in dependencies count).
